extensions-contrib/gce-extensions/src/main/java/org/apache/druid/indexing/overlord/autoscaling/gce/GceAutoScaler.java

```diff
@@ -0,0 +1,539 @@
+      Thread.sleep(POLL_INTERVAL_MS);
+      Compute.ZoneOperations.Get get = compute.zoneOperations().get(
+          envConfig.getProjectId(),
+          envConfig.getZoneName(),
+          opId
+      );
+      operation = get.execute();
+      if (operation != null) {
+        status = operation.getStatus();
+      }
+    }
+    throw new InterruptedException(
+        StringUtils.format("Timed out waiting for operation %s to complete", opId)
+    );
+  }
+
+  /**
+   * When called resizes envConfig.getManagedInstanceGroupName() increasing it by creating
+   * envConfig.getNumInstances() new workers (unless the maximum is reached). Return the
+   * IDs of the workers created
+   */
+  @Override
+  public AutoScalingData provision()
+  {
+    final String project = envConfig.getProjectId();
+    final String zone = envConfig.getZoneName();
+    final int numInstances = envConfig.getNumInstances();
+    final String managedInstanceGroupName = envConfig.getManagedInstanceGroupName();
+
+    try {
+      List<String> before = getRunningInstances();
+      log.debug("Existing instances [%s]", String.join(",", before));
+
+      int toSize = Math.min(before.size() + numInstances, getMaxNumWorkers());
+      if (before.size() >= toSize) {
+        // nothing to scale
+        return new AutoScalingData(new ArrayList<>());
+      }
+      log.info("Asked to provision instances, will resize to %d", toSize);
+
+      Compute computeService = createComputeService();
+      Compute.InstanceGroupManagers.Resize request =
+              computeService.instanceGroupManagers().resize(project, zone,
+                      managedInstanceGroupName, toSize);
+
+      Operation response = request.execute();
+      Operation.Error err = waitForOperationEnd(computeService, response);
+      if (err == null || err.isEmpty()) {
+        List<String> after = null;
+        // as the waitForOperationEnd only waits for the operation to be scheduled
+        // this loop waits until the requested machines actually go up (or up to a
+        // certain amount of retries in checking)
+        for (int i = 0; i < RUNNING_INSTANCES_MAX_RETRIES; i++) {
+          after = getRunningInstances();
+          if (after.size() == toSize) {
+            break;
+          }
+          log.info("Machines not up yet, waiting");
+          Thread.sleep(POLL_INTERVAL_MS);
+        }
+        after.removeAll(before); // these should be the new ones
+        log.info("Added instances [%s]", String.join(",", after));
+        return new AutoScalingData(after);
+      } else {
+        log.error("Unable to provision instances: %s", err.toPrettyString());
+      }
+    }
+    catch (Exception e) {
+      log.error(e, "Unable to provision any gce instances.");
+    }
+
+    return new AutoScalingData(new ArrayList<>());
+  }
+
+  /**
+   * Terminates the instances in the list of IPs provided by the caller
+   */
+  @Override
+  public AutoScalingData terminate(List<String> ips)
```

Comment:
done.  this made me found a small problem in the existing code as it was calling id → ip before returning which would result in an empty collections.  not a big deal given how the caller uses that information and also because we already waited for the instances to be down anyway before returning.